core/server/master/src/main/java/alluxio/master/block/DefaultBlockMaster.java

```diff
@@ -566,53 +592,29 @@ public void commitBlock(long workerId, long usedBytesOnTier, String tierAlias, l
     // Lock the worker metadata first.
     try (JournalContext journalContext = createJournalContext()) {
       synchronized (worker) {
-        // Loop until block metadata is successfully locked.
-        while (true) {
-          boolean newBlock = false;
-          MasterBlockInfo block = mBlocks.get(blockId);
-          if (block == null) {
-            // The block metadata doesn't exist yet.
-            block = new MasterBlockInfo(blockId, length);
-            newBlock = true;
+        try (LockResource lr = lockBlock(blockId)) {
+          Optional<BlockMeta> block = mBlockStore.getBlock(blockId);
+          if (!block.isPresent() || block.get().getLength() != length) {
```

Comment:
how about the part handling `block.getLength() == Constants.UNKNOWN_SIZE) `? it seems this part of code is removed

Reply:
updated to add a warning and not change the length if we try to change the length of a block that already has its length set. I don't think it should be possible, but it's good to get a warning in the logs in case it does happen.

```java
if (block.isPresent() && block.get().getLength() != Constants.UNKNOWN_SIZE) {
  LOG.warn("Rejecting attempt to change block length from {} to {}",
  block.get().getLength(), length);
} else {
  mBlockStore.putBlock(blockId, BlockMeta.newBuilder().setLength(length).build());
  BlockInfoEntry blockInfo =
  BlockInfoEntry.newBuilder().setBlockId(blockId).setLength(length).build();
  journalContext.append(JournalEntry.newBuilder().setBlockInfo(blockInfo).build());
}
```
@gpang can you confirm that it's a bug if the length gets changed from a value other than `Constants.UNKNOWN_SIZE`?

Reply:
@aaudiber yeah, the block size should not be modified after it is a "known" size.